src/stdlib/io.rs

```diff
@@ -69,4 +88,831 @@ pub fn load_io<'gc>(ctx: Context<'gc>) {
+                        Ok(CallbackReturn::Return)
+                    } else {
+                        Err("expected `file`".into_value(ctx).into())
+                    }
+                }
+                _ => Err("expected `string` or `file`".into_value(ctx).into()),
+            }
+        }),
+    );
+
+    io.set_field(
+        ctx,
+        "close",
+        Callback::from_fn(&ctx, |ctx, _, mut stack| {
+            if stack.is_empty() {
+                let output = IO_STATE.with(|state| state.output());
+
+                if !output.is_std() {
+                    if let Err(err) = output.close() {
+                        stack.replace(
+                            ctx,
+                            (Value::Nil, err.to_string(), err.raw_os_error().unwrap_or(0)),
+                        );
+                        return Ok(CallbackReturn::Return);
+                    }
+                    IO_STATE.with(|state| state.replace_output(IoFile::stdout()));
+                    stack.replace(ctx, true);
+                    return Ok(CallbackReturn::Return);
+                }
+
+                stack.replace(ctx, true);
+                return Ok(CallbackReturn::Return);
+            }
+
+            let file: UserData = stack.consume(ctx)?;
+            let file_wrapper = if let Ok(fw) = file.downcast_static::<IoFile>() {
+                fw
+            } else {
+                return Err("expected `file`".into_value(ctx).into());
+            };
+
+            match file_wrapper.close() {
+                Ok(_) => {
+                    stack.replace(ctx, true);
+                    Ok(CallbackReturn::Return)
+                }
+                Err(e) => {
+                    stack.replace(
+                        ctx,
+                        (Value::Nil, e.to_string(), e.raw_os_error().unwrap_or(0)),
+                    );
+                    Ok(CallbackReturn::Return)
+                }
+            }
+        }),
+    );
+
+    io.set_field(
+        ctx,
+        "flush",
+        Callback::from_fn(&ctx, |ctx, _, mut stack| {
+            let output = IO_STATE.with(|state| state.output());
+
+            if let Err(err) = output.flush() {
+                stack.replace(
+                    ctx,
+                    (Value::Nil, err.to_string(), err.raw_os_error().unwrap_or(0)),
+                );
+                Ok(CallbackReturn::Return)
+            } else {
+                stack.replace(ctx, UserData::new_static(&ctx, output.clone()));
+                Ok(CallbackReturn::Return)
+            }
+        }),
+    );
+
+    io.set_field(
+        ctx,
+        "read",
+        Callback::from_fn(&ctx, |ctx, _, mut stack| {
+            let input = IO_STATE.with(|state| state.input());
+
+            let file: Table = ctx.io_metatable().get(ctx, MetaMethod::Index)?;
+            let read: Callback = file.get(ctx, "read")?;
+
+            stack.into_front(ctx, Value::UserData(UserData::new_static(&ctx, input)));
```

Comment:
Like `open`: `read`, `write` and `lines` should be implemented as a Rust function or closure that can be directly called here, rather than having to go through the global lookup/userdata/call setup.